gprofiler/profilers/java.py

```diff
@@ -665,10 +654,30 @@ def _is_profiling_supported(self, process: Process) -> bool:
 
         return True
 
+    def _check_async_profiler_loaded(self, process: Process) -> bool:
+        if JavaSafemodeOptions.AP_LOADED_CHECK not in self._java_safemode:
+            # don't care
+            return False
+
+        for mmap in process.memory_maps():
+            if "libasyncProfiler.so" in mmap.path and not mmap.path.startswith(TEMPORARY_STORAGE_PATH):
+                logger.warning(
+                    "Non-gProfiler async-profiler is already loaded to the target process."
+                    " Disable --java-safemode=ap-loaded-check to bypass this check.",
+                    pid=process.pid,
+                    ap_path=mmap.path,
+                )
+                return True
+
+        return False
+
     def _profile_process(self, process: Process) -> Optional[StackToSampleCount]:
         if not self._is_profiling_supported(process):
             return None
 
+        if self._check_async_profiler_loaded(process):
```

Comment:
Is there a special reason you moved it to `JavaProfiler` from `AsyncProfiledProcess`?

Reply:
I preferred the logging line + `return None` here approach, to the exception-based approach. In `AsyncProfiledProcess` I currently lack the API for the `return None`, and we already do some "compatibility" checks here, so this seemed like the right spot.

Reply:
Makes sense